File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> comp;
        comp[')'] = '(';
        comp[']'] = '[';
        comp['}'] = '{';
        stack<char> st;

        for (char ch : s) {
            if (ch == '(' || ch == '[' || ch == '{') {
                st.push(ch);
            } else {
                if (st.size() == 0 || st.top() != comp[ch]) {
                    return false;
                }
                st.pop();
            }
        }

        if (st.size() > 0) {
            return false;
        }
        
        return true;
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        Map<Character, Character> comp = new HashMap<>();
        comp.put(')', '(');
        comp.put('}', '{');
        comp.put(']', '[');

        for (char ch : s.toCharArray()) {
            if (ch == '(' || ch == '{' || ch == '[') {
                st.push(ch);
            } else {
                if (st.size() == 0) {
                    return false;
                }

                if (st.peek() == comp.get(ch)) {
                    st.pop();
                } else {
                    return false;
                }
            }
        }

        return true;
    }
}